File: nodes/core/market/filters/duplicate_symbol_filter_node.py
import logging
from typing import Any

from core.types_registry import AssetSymbol, NodeCategory, OHLCVBar, get_type
from nodes.base.base_node import Base

logger = logging.getLogger(__name__)


class DuplicateSymbolFilter(Base):
    """
    Compare symbols across up to 3 inputs to find common or unique symbols.
    
    Accepts OHLCV bundles or symbol lists from ExtractSymbols.
    Connect up to 3 inputs to compare symbols between them.
    """

    CATEGORY = NodeCategory.MARKET

    inputs = {
        "input_1": (get_type("OHLCVBundle") | get_type("AssetSymbolList")) | None,
        "input_2": (get_type("OHLCVBundle") | get_type("AssetSymbolList")) | None,
        "input_3": (get_type("OHLCVBundle") | get_type("AssetSymbolList")) | None,
    }

    optional_inputs = ["input_1", "input_2", "input_3"]

    outputs = {
        "filtered_ohlcv_bundle": get_type("OHLCVBundle"),
    }

    default_params = {
        "operation": "common",  # common, unique_to_1, unique_to_2, unique_to_3, all
        "compare_by": "ticker",  # ticker, symbol_string
        "case_insensitive": True,
    }

    params_meta = [
        {
            "name": "operation",
            "type": "combo",
            "default": "common",
            "options": [
                "common",
                "unique_to_1",
                "unique_to_2",
                "unique_to_3",
                "all",
            ],
            "label": "Operation",
            "description": "common: symbols in all connected inputs | unique_to_X: symbols only in input_X | all: union of all symbols",
        },
        {
            "name": "compare_by",
            "type": "combo",
            "default": "ticker",
            "options": ["ticker", "symbol_string"],
            "label": "Compare By",
            "description": "Compare symbols by ticker (base symbol) or full symbol string",
        },
        {
            "name": "case_insensitive",
            "type": "boolean",
            "default": True,
            "label": "Case Insensitive",
            "description": "Compare symbols case-insensitively",
        },
    ]

    def __init__(
        self,
        id: int,
        params: dict[str, Any] | None = None,
        graph_context: dict[str, Any] | None = None,
    ):
        super().__init__(id, params, graph_context)

    def _normalize_input(self, input_data: Any) -> dict[AssetSymbol, list[OHLCVBar]]:
        """Normalize input to OHLCV bundle format.
        
        Handles both OHLCV bundles (dict[AssetSymbol, list[OHLCVBar]]) 
        and symbol lists (list[AssetSymbol]).
        """
        if input_data is None:
            return {}
        
        # If it's a list, treat as symbol list and create empty bundles
        if isinstance(input_data, list):
            return {symbol: [] for symbol in input_data if isinstance(symbol, AssetSymbol)}
        
        # If it's a dict, assume it's an OHLCV bundle
        if isinstance(input_data, dict):
            return {k: v for k, v in input_data.items() if isinstance(k, AssetSymbol)}
        
        return {}

    def _get_symbols_from_input(self, input_data: dict[AssetSymbol, list[OHLCVBar]]) -> list[AssetSymbol]:
        """Extract list of symbols from normalized input."""
        return list(input_data.keys())

    def _make_key(self, symbol: AssetSymbol) -> str:
        """Create a comparison key from a symbol based on compare_by setting."""
        compare_by = self.params.get("compare_by", "ticker")
        case_insensitive = self.params.get("case_insensitive", True)

        if compare_by == "ticker":
            key = symbol.ticker
        else:  # symbol_string
            key = str(symbol)

        if case_insensitive:
            return key.lower()
        return key

    async def _execute_impl(self, inputs: dict[str, Any]) -> dict[str, Any]:
        # Normalize all inputs to OHLCV bundle format
        input_1 = self._normalize_input(inputs.get("input_1"))
        input_2 = self._normalize_input(inputs.get("input_2"))
        input_3 = self._normalize_input(inputs.get("input_3"))

        # Get operation mode
        operation = self.params.get("operation", "common")

        # Determine which inputs are connected
        connected_inputs = []
        if input_1:
            connected_inputs.append(("input_1", input_1))
        if input_2:
            connected_inputs.append(("input_2", input_2))
        if input_3:
            connected_inputs.append(("input_3", input_3))

        if not connected_inputs:
            logger.warning("DuplicateSymbolFilter: No inputs connected")
            return {"filtered_ohlcv_bundle": {}}

        # Build symbol sets for each input
        symbol_sets: dict[str, set[str]] = {}
        symbol_maps: dict[str, dict[str, AssetSymbol]] = {}
        input_bundles: dict[str, dict[AssetSymbol, list[OHLCVBar]]] = {}

        for name, bundle in connected_inputs:
            symbols = self._get_symbols_from_input(bundle)
            keys = set()
            key_to_symbol: dict[str, AssetSymbol] = {}
            
            for symbol in symbols:
                k = self._make_key(symbol)
                keys.add(k)
                if k not in key_to_symbol:
                    key_to_symbol[k] = symbol
            
            symbol_sets[name] = keys
            symbol_maps[name] = key_to_symbol
            input_bundles[name] = bundle

        # Execute operation
        filtered: dict[AssetSymbol, list[OHLCVBar]] = {}

        if operation == "common":
            # Symbols in ALL connected inputs
            if len(connected_inputs) == 0:
                pass
            elif len(connected_inputs) == 1:
                # If only one input, return all symbols from it
                name, bundle = connected_inputs[0]
                filtered = bundle
            else:
                # Intersection of all inputs
                common_keys = symbol_sets[connected_inputs[0][0]]
                for name, _ in connected_inputs[1:]:
                    common_keys &= symbol_sets[name]
                
                # Use first input's bundle for data
                first_name, first_bundle = connected_inputs[0]
                for k in common_keys:
                    symbol = symbol_maps[first_name][k]
                    # Try to get data from any input that has it
                    for name, bundle in connected_inputs:
                        if symbol in bundle:
                            filtered[symbol] = bundle[symbol]
                            break
                    else:
                        # If no bundle has data, create empty entry
                        filtered[symbol] = []

        elif operation == "unique_to_1" and "input_1" in symbol_sets:
            # Symbols only in input_1, not in others
            unique_keys = symbol_sets["input_1"]
            for name, _ in connected_inputs:
                if name != "input_1":
                    unique_keys -= symbol_sets[name]
            
            for k in unique_keys:
                symbol = symbol_maps["input_1"][k]
                filtered[symbol] = input_bundles["input_1"].get(symbol, [])

        elif operation == "unique_to_2" and "input_2" in symbol_sets:
            # Symbols only in input_2, not in others
            unique_keys = symbol_sets["input_2"]
            for name, _ in connected_inputs:
                if name != "input_2":
                    unique_keys -= symbol_sets[name]
            
            for k in unique_keys:
                symbol = symbol_maps["input_2"][k]
                filtered[symbol] = input_bundles["input_2"].get(symbol, [])

        elif operation == "unique_to_3" and "input_3" in symbol_sets:
            # Symbols only in input_3, not in others
            unique_keys = symbol_sets["input_3"]
            for name, _ in connected_inputs:
                if name != "input_3":
                    unique_keys -= symbol_sets[name]
            
            for k in unique_keys:
                symbol = symbol_maps["input_3"][k]
                filtered[symbol] = input_bundles["input_3"].get(symbol, [])

        elif operation == "all":
            # Union of all symbols from all inputs
            all_keys: set[str] = set()
            for name, _ in connected_inputs:
                all_keys |= symbol_sets[name]
            
            # Build merged result, preferring data from first input that has it
            for k in all_keys:
                symbol = None
                bundle_data = []
                
                # Find symbol and data from first input that has this key
                for name, bundle in connected_inputs:
                    if k in symbol_maps[name]:
                        symbol = symbol_maps[name][k]
                        if symbol in bundle:
                            bundle_data = bundle[symbol]
                        break
                
                if symbol:
                    filtered[symbol] = bundle_data

        logger.info(f"DuplicateSymbolFilter: Operation '{operation}' returned {len(filtered)} symbols")
        return {"filtered_ohlcv_bundle": filtered}

File: nodes/market/filters/duplicate-symbol-filter-node.ts
// src/nodes/core/market/filters/duplicate-symbol-filter-node.ts
// Translated from: nodes/core/market/filters/duplicate_symbol_filter_node.py

import { BaseFilter } from './base/base-filter-node';
import { getType } from '@fig-node/core';
import type {
  NodeInputs,
  NodeOutputs,
  ParamMeta,
  DefaultParams,
  OHLCVBar,
  OHLCVBundle,
  NodeUIConfig,
} from '@fig-node/core';
import { AssetSymbol } from '@fig-node/core';

/**
 * Compare symbols across up to 3 inputs to find common or unique symbols.
 *
 * Accepts OHLCV bundles or symbol lists from ExtractSymbols.
 * Connect up to 3 inputs to compare symbols between them.
 */
export class DuplicateSymbolFilter extends BaseFilter {
  static override inputs: Record<string, unknown> = {
    ohlcv_bundle_1: getType('OHLCVBundle'),
    ohlcv_bundle_2: getType('OHLCVBundle'),
    ohlcv_bundle_3: getType('OHLCVBundle'),
  };

  static override defaultParams: DefaultParams = {
    operation: 'common', // common, unique_to_1, unique_to_2, unique_to_3, all
    compare_by: 'ticker', // ticker, symbol_string
    case_insensitive: true,
  };

  static override paramsMeta: ParamMeta[] = [
    {
      name: 'operation',
      type: 'combo',
      default: 'common',
      options: ['common', 'unique_to_1', 'unique_to_2', 'unique_to_3', 'all'],
      label: 'Operation',
      description:
        'common: symbols in all connected inputs | unique_to_X: symbols only in ohlcv_bundle_X | all: union of all symbols',
    },
    {
      name: 'compare_by',
      type: 'combo',
      default: 'ticker',
      options: ['ticker', 'symbol_string'],
      label: 'Compare By',
      description: 'Compare symbols by ticker (base symbol) or full symbol string',
    },
    {
      name: 'case_insensitive',
      type: 'boolean',
      default: true,
      label: 'Case Insensitive',
      description: 'Compare symbols case-insensitively',
    },
  ];

  static uiConfig: NodeUIConfig = {
    size: [220, 100],
    displayResults: false,
    resizable: false,
  };

  /**
   * Normalize input to OHLCV bundle format.
   * Handles both OHLCV bundles (Map<AssetSymbol, OHLCVBar[]>) and symbol lists (AssetSymbol[]).
   */
  private normalizeInput(inputData: unknown): OHLCVBundle {
    if (inputData === null || inputData === undefined) {
      return new Map();
    }

    // If it's already a Map
    if (inputData instanceof Map) {
      const result: OHLCVBundle = new Map();
      for (const [key, value] of inputData) {
        if (key instanceof AssetSymbol && Array.isArray(value)) {
          result.set(key, value as OHLCVBar[]);
        }
      }
      return result;
    }

    // If it's an array, treat as symbol list and create empty bundles
    if (Array.isArray(inputData)) {
      const result: OHLCVBundle = new Map();
      for (const item of inputData) {
        if (item instanceof AssetSymbol) {
          result.set(item, []);
        }
      }
      return result;
    }

    // If it's a plain object, try to convert
    if (typeof inputData === 'object') {
      const result: OHLCVBundle = new Map();
      // This would need proper deserialization in production
      return result;
    }

    return new Map();
  }

  /**
   * Extract list of symbols from normalized input.
   */
  private getSymbolsFromInput(inputData: OHLCVBundle): AssetSymbol[] {
    return Array.from(inputData.keys());
  }

  /**
   * Create a comparison key from a symbol based on compare_by setting.
   */
  private makeKey(symbol: AssetSymbol): string {
    const compareBy = this.params.compare_by ?? 'ticker';
    const caseInsensitive = this.params.case_insensitive ?? true;

    let key: string;
    if (compareBy === 'ticker') {
      key = symbol.ticker;
    } else {
      // symbol_string
      key = symbol.toString();
    }

    if (caseInsensitive) {
      return key.toLowerCase();
    }
    return key;
  }

  protected override async executeImpl(inputs: NodeInputs): Promise<NodeOutputs> {
    // Normalize all inputs to OHLCV bundle format
    const ohlcvBundle1 = this.normalizeInput(inputs.ohlcv_bundle_1);
    const ohlcvBundle2 = this.normalizeInput(inputs.ohlcv_bundle_2);
    const ohlcvBundle3 = this.normalizeInput(inputs.ohlcv_bundle_3);

    // Get operation mode
    const operation = this.params.operation ?? 'common';

    // Determine which inputs are connected
    const connectedInputs: Array<[string, OHLCVBundle]> = [];
    if (ohlcvBundle1.size > 0) {
      connectedInputs.push(['ohlcv_bundle_1', ohlcvBundle1]);
    }
    if (ohlcvBundle2.size > 0) {
      connectedInputs.push(['ohlcv_bundle_2', ohlcvBundle2]);
    }
    if (ohlcvBundle3.size > 0) {
      connectedInputs.push(['ohlcv_bundle_3', ohlcvBundle3]);
    }

    if (connectedInputs.length === 0) {
      console.warn('DuplicateSymbolFilter: No inputs connected');
      return { filtered_ohlcv_bundle: new Map() };
    }

    // Build symbol sets for each input
    const symbolSets: Map<string, Set<string>> = new Map();
    const symbolMaps: Map<string, Map<string, AssetSymbol>> = new Map();
    const inputBundles: Map<string, OHLCVBundle> = new Map();

    for (const [name, bundle] of connectedInputs) {
      const symbols = this.getSymbolsFromInput(bundle);
      const keys: Set<string> = new Set();
      const keyToSymbol: Map<string, AssetSymbol> = new Map();

      for (const symbol of symbols) {
        const k = this.makeKey(symbol);
        keys.add(k);
        if (!keyToSymbol.has(k)) {
          keyToSymbol.set(k, symbol);
        }
      }

      symbolSets.set(name, keys);
      symbolMaps.set(name, keyToSymbol);
      inputBundles.set(name, bundle);
    }

    // Execute operation
    const filtered: OHLCVBundle = new Map();

    if (operation === 'common') {
      // Symbols in ALL connected inputs
      if (connectedInputs.length === 0) {
        // No-op
      } else if (connectedInputs.length === 1) {
        // If only one input, return all symbols from it
        const [, bundle] = connectedInputs[0];
        for (const [symbol, data] of bundle) {
          filtered.set(symbol, data);
        }
      } else {
        // Intersection of all inputs
        let commonKeys = new Set(symbolSets.get(connectedInputs[0][0])!);
        for (let i = 1; i < connectedInputs.length; i++) {
          const [name] = connectedInputs[i];
          const otherKeys = symbolSets.get(name)!;
          commonKeys = new Set([...commonKeys].filter((k) => otherKeys.has(k)));
        }

        // Use first input's bundle for data
        const firstName = connectedInputs[0][0];
        for (const k of commonKeys) {
          const symbol = symbolMaps.get(firstName)!.get(k)!;
          // Try to get data from any input that has it
          let found = false;
          for (const [_name, bundle] of connectedInputs) {
            if (bundle.has(symbol)) {
              filtered.set(symbol, bundle.get(symbol)!);
              found = true;
              break;
            }
          }
          if (!found) {
            // If no bundle has data, create empty entry
            filtered.set(symbol, []);
          }
        }
      }
    } else if (operation === 'unique_to_1' && symbolSets.has('ohlcv_bundle_1')) {
      // Symbols only in ohlcv_bundle_1, not in others
      let uniqueKeys = new Set(symbolSets.get('ohlcv_bundle_1')!);
      for (const [name] of connectedInputs) {
        if (name !== 'ohlcv_bundle_1') {
          const otherKeys = symbolSets.get(name)!;
          uniqueKeys = new Set([...uniqueKeys].filter((k) => !otherKeys.has(k)));
        }
      }

      for (const k of uniqueKeys) {
        const symbol = symbolMaps.get('ohlcv_bundle_1')!.get(k)!;
        filtered.set(symbol, inputBundles.get('ohlcv_bundle_1')!.get(symbol) ?? []);
      }
    } else if (operation === 'unique_to_2' && symbolSets.has('ohlcv_bundle_2')) {
      // Symbols only in ohlcv_bundle_2, not in others
      let uniqueKeys = new Set(symbolSets.get('ohlcv_bundle_2')!);
      for (const [name] of connectedInputs) {
        if (name !== 'ohlcv_bundle_2') {
          const otherKeys = symbolSets.get(name)!;
          uniqueKeys = new Set([...uniqueKeys].filter((k) => !otherKeys.has(k)));
        }
      }

      for (const k of uniqueKeys) {
        const symbol = symbolMaps.get('ohlcv_bundle_2')!.get(k)!;
        filtered.set(symbol, inputBundles.get('ohlcv_bundle_2')!.get(symbol) ?? []);
      }
    } else if (operation === 'unique_to_3' && symbolSets.has('ohlcv_bundle_3')) {
      // Symbols only in ohlcv_bundle_3, not in others
      let uniqueKeys = new Set(symbolSets.get('ohlcv_bundle_3')!);
      for (const [name] of connectedInputs) {
        if (name !== 'ohlcv_bundle_3') {
          const otherKeys = symbolSets.get(name)!;
          uniqueKeys = new Set([...uniqueKeys].filter((k) => !otherKeys.has(k)));
        }
      }

      for (const k of uniqueKeys) {
        const symbol = symbolMaps.get('ohlcv_bundle_3')!.get(k)!;
        filtered.set(symbol, inputBundles.get('ohlcv_bundle_3')!.get(symbol) ?? []);
      }
    } else if (operation === 'all') {
      // Union of all symbols from all inputs
      const allKeys: Set<string> = new Set();
      for (const [name] of connectedInputs) {
        for (const k of symbolSets.get(name)!) {
          allKeys.add(k);
        }
      }

      // Build merged result, preferring data from first input that has it
      for (const k of allKeys) {
        let symbol: AssetSymbol | null = null;
        let bundleData: OHLCVBar[] = [];

        // Find symbol and data from first input that has this key
        for (const [name, bundle] of connectedInputs) {
          if (symbolMaps.get(name)!.has(k)) {
            symbol = symbolMaps.get(name)!.get(k)!;
            if (bundle.has(symbol)) {
              bundleData = bundle.get(symbol)!;
            }
            break;
          }
        }

        if (symbol) {
          filtered.set(symbol, bundleData);
        }
      }
    }

    console.log(
      `DuplicateSymbolFilter: Operation '${operation}' returned ${filtered.size} symbols`
    );
    return { filtered_ohlcv_bundle: filtered };
  }
}
